Update README and release-notes script (#23)
kubevirt/kubevirt-velero-plugin: README.md

# Kubevirt Velero Plugin

> This repository contains Velero plugins. Thanks to this plugin, velero can correctly backup and restore 

Reviewer: ```suggestion
This repository contains Velero plugins. Thanks to this plugin, [Velero](https://velero.io/) can correctly backup and restore 
```

> VMs, DVs and other resources managed by kubevirt - https://kubevirt.io/. 

Reviewer: ```suggestion
VirtualMachines, DataVolumess and other resources managed by [KubeVirt](https://kubevirt.io/). 
```

For more information on Velero check https://velero.io/.

This is still a work in progress, not intended for production use.

## Install

> To install remove the plugin check current velero documentation https://velero.io/docs/v1.7/overview-plugins/.

Reviewer: ```suggestion
To install the plugin check current velero documentation https://velero.io/docs/v1.7/overview-plugins/.
```

Below example for kubevirt-velero-plugin version v0.2.0 on Velero 1.7.0

```bash
velero plugin add quay.io/kubevirt/kubevirt-velero-plugin:v0.2.0
```
## Compatibility 

> Plugin versions and respective Velero/Kubevirt/CDI versions that are tested to be compatible.

Reviewer: Please link on the first use of the term CDI as it is not known by most readers.

Author: done

| Plugin Version  | Velero Version | Kubevirt Version | CDI Version   |
|-----------------|----------------|------------------|---------------|
| v0.2.0          | v1.6.x, v1.7.x | v0.48.x          | \>= v1.37.0 |

> 

Reviewer: Can you add a usage section? It may repeat a line or two from Velero docs (I like what you did earlier, qualifying the example for a specific Velero version)

Author: Added usage with example. I may add some more detailed readme inside `example` directory.

# Kubevirt Velero Plugin

This repository contains Velero plugins. Thanks to this plugin, [Velero](https://velero.io/) can correctly backup and restore
VirtualMachines, DataVolumess and other resources managed by [KubeVirt](https://kubevirt.io/) and 
[CDI](https://github.com/kubevirt/containerized-data-importer/).

For more information on Velero check https://velero.io/.

This is still a work in progress, not intended for production use.

## Kinds of Plugins Included

Kubevirt Velero Plugin implements two kinds of plugins:

> - **Backup Item Action** - performs arbitrary logic on individual items prior to storing them in the backup file.

Author: That is right and I plan on describing all of this the same way it is described in velero-plugin-for-csi:
https://github.com/vmware-tanzu/velero-plugin-for-csi#kinds-of-plugins-included

But I will not finish it today. That can be done in January. Here is the issue: https://github.com/kubevirt/kubevirt-velero-plugin/issues/24

> - **Backup Item Action** - performs arbitrary logic on individual items prior to storing them in the backup file.

Author: I have now added better plugin description.

As for the pvc backup, this is all done by velero, the plugin does not change anything with the handling of PVC contents. 

The velero either creates a snapshot to save PVC contents or makes a file copy with the use of restic.

The plugin operates on following resources: DataVolume, PersistentVolumeClaim, Pod, VirtualMachine, VirtualMachineInstance
 
## Compatibility

Plugin versions and respective Velero/Kubevirt/CDI versions that are tested to be compatible.

| Plugin Version  | Velero Version | Kubevirt Version | CDI Version  |
|-----------------|----------------|------------------|--------------|
| v0.2.0          | v1.6.x, v1.7.x | v0.48.x          | \>= v1.37.0  |

## Install

To install the plugin check current velero documentation https://velero.io/docs/v1.7/overview-plugins/.
Below example for kubevirt-velero-plugin version v0.2.0 on Velero 1.7.0

```bash
velero plugin add quay.io/kubevirt/kubevirt-velero-plugin:v0.2.0
```

## Backup/Restore Virtual Machines Using the Plugin

When the plugin is deployed it is already effective. There is nothing special required to make it work.
### 1. Create a Virtual Machine

```bash
kubectl create namespace demo
kubectl create -f example/datavolume.yaml -n demo
kubectl create -f example/vm.yaml -n demo
```
Start a VM (wait for condition AgentConnected):
```bash
virtctl start example-vm -n demo`

> ```

Reviewer: you can make the example shorter if you add `  running: true` to the vm.yaml (not sure you should care about this)

Author: sure, that simplifies instruction

### **VMBackupItemAction** 
A plugin that backs up the `VirtualMachine`
 
The plugin checks if a `VM` can be safely backed up and if the backup contains all required objects for the successful restore. 
The plugin also returns the underlying `DataVolume` if a VM has `DataVolumeTemplate` and `virtualmachineinstances` as extra items to back up.

### **VMIBackupItemAction** 
A plugin that backs up the `VirtualMachineInstance`
 
The plugin checks if a `VMI` can be safely backed up and if the backup contains all required objects for the successful restore.
The plugin also returns the underlying VM volumes (`DataVolume` and `PersistentVolumeClaim`) and launcher `pod` as extra items to back up.

### **VMRestoreItemAction**
A plugin that restores the `VirtualMachine`
 
Adds a `datavolumes` to list of restored items.

### **VMIRestoreItemAction** 
A plugin that restores the `VirtualMachineInstance`

Skips the VMI if owned by a VM. The plugin also clears restricted labels, so the VMI is not rejected by kubevirt.  The restricted labels contain runtime information about the underlying KVM object.

### **PodRestoreItemAction**

> A plugin that restores the virt-launcher `Pod`

Reviewer: I'm surprised to see it. Restoring VMI is not enough?

Author: You are right, now when I checked the code again I see that the plugin just makes sure launcher pod is *NOT* restored:
"Launcher Pod should be unconditionally skipped"
